Add ML-DSA post-quantum signatures to `_CryptoExtras`

fpseverino opened the pull request:
Add support for ML-DSA post-quantum digital signatures inside `_CryptoExtras`.

<!-- Thanks for contributing to Swift Crypto! Before you submit your issue, please make sure you followed our checklist and check the appropriate boxes by putting an x in the [ ]: [x] -->

### Checklist
- [x] I've run tests to see all new and existing tests pass
- [x] I've followed the code style of the rest of the project
- [x] I've read the [Contribution Guidelines](CONTRIBUTING.md)
- [x] I've updated the documentation if necessary

#### If you've made changes to `gyb` files
- [ ] I've run `.script/generate_boilerplate_files_with_gyb` and included updated generated files in a commit of this pull request

### Motivation:

With the advent of quantum computing, the mathematical foundations on which the cryptographic protocols in use today are based have been questioned, as they can easily be circumvented and violated by quantum computers.

While waiting for the creation of quantum computers that work at full capacity, and to protect network communications from "[Harvest Now, Decrypt Later](https://en.wikipedia.org/wiki/Harvest_now,_decrypt_later)" attacks, the cryptographic community is working on post-quantum cryptography algorithms, which work on the traditional computers we use today, but are resistant to future attacks by quantum computers.

One of these algorithms is ML-DSA (AKA Dilithium), a module lattice-based signature scheme standardized by NIST in [FIPS 204](https://csrc.nist.gov/pubs/fips/204/final), that is available inside BoringSSL.

By including ML-DSA inside Swift Crypto, we can get closer to normalizing quantum secure algorithms and start implementing them into our apps and libraries to make them quantum-proof.

### Modifications:

Added a `MLDSA65` enum inside the `_CryptoExtras` module with corresponding `PrivateKey`, `PublicKey` and `Signature` structs that use BoringSSL methods to produce and verify ML-DSA-65 digital signatures, with the code style of other signature schemes in the library.

Added tests that cover use cases of the ML-DSA scheme, including [test vectors taken from the BoringSSL repo](https://boringssl.googlesource.com/boringssl/+/refs/heads/master/crypto/mldsa/mldsa_nist_keygen_tests.txt) (extracted from a `.txt` file and encoded in JSON).

### Result:

ML-DSA-65 digital signatures can be created and verified with Swift Crypto.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,319 @@
+extension MLDSA {
+    /// A ML-DSA-65 private key.
+    public struct PrivateKey: Sendable {
+        fileprivate let pointer: UnsafeMutablePointer<MLDSA65_private_key>
```
This is unsound. It's never possible for a copyable `struct` to own an unsafe pointer safely. This needs to have a backing class to own the pointer.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,319 @@
+            }
+
+            let seedPtr = UnsafeMutablePointer<UInt8>.allocate(capacity: MLDSA.seedSizeInBytes)
+            seedPtr.initialize(from: seed.regions.flatMap { $0 }, count: MLDSA.seedSizeInBytes)
```
There's a [`copyBytes(to:)`](https://developer.apple.com/documentation/foundation/dataprotocol/3139158-copybytes) function we can use for this.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,319 @@
+        }
+
+        /// The size of the public key in bytes.
+        public static let bytesCount = 1952
```
We should avoid making this public.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,319 @@
+        /// - Parameter derRepresentation: The DER representation of the private key.
+        /// 
+        /// - Throws: `CryptoKitError.incorrectKeySize` if the DER representation is not the correct size or ``CryptoMLDSAError/keyGenerationFailure`` if the key could not be generated.
+        public init(derRepresentation: some DataProtocol) throws {
```
This isn't parsing the DER representation, this is closer to the `rawRepresentation`.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,319 @@
+        /// Initialize a ML-DSA-65 private key from a PEM representation.
+        /// 
+        /// - Parameter pemRepresentation: The PEM representation of the private key.
+        public init(pemRepresentation: String) throws {
```
Similarly, there's no PEM here.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,319 @@
+
+        /// The public key associated with this private key.
+        public var publicKey: PublicKey {
+            get throws {
```
I don't think there's any reason for this to throw: we should always be able to produce the public counterpart for a private key.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,319 @@
+extension MLDSA {
+    /// A ML-DSA-65 public key.
+    public struct PublicKey: Sendable {
+        private let pointer: UnsafeMutablePointer<MLDSA65_public_key>
```
Same note on this pointer.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,319 @@
+        }
+
+        /// The size of the private key in bytes.
+        public static let bytesCount = 4032
```
Let's not offer this API publicly.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,319 @@
+        public func signature(for data: some DataProtocol, context: [UInt8]? = nil) throws -> Signature {
+            let output = try Array<UInt8>(unsafeUninitializedCapacity: Signature.bytesCount) { bufferPtr, length in
+                let result = data.regions.first!.withUnsafeBytes { dataPtr in
+                    let contextPointer = context?.withUnsafeBufferPointer { $0.baseAddress }
```
This is unsound: you must not escape this pointer from the closure.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,319 @@
+        /// Any additional bytes in the seed are ignored.
+        /// 
+        /// - Parameter seed: The seed to use to generate the public key.
+        public init(from seed: some DataProtocol) throws {
```
I don't think this API makes sense: users should construct the private key, not the public one.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,319 @@
+        /// - Parameter derRepresentation: The DER representation of the public key.
+        /// 
+        /// - Throws: `CryptoKitError.incorrectKeySize` if the DER representation is not the correct size or ``CryptoMLDSAError/keyGenerationFailure`` if the key could not be generated.
+        public init(derRepresentation: some DataProtocol) throws {
```
This isn't parsing the DER format, it's more like the raw format.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,319 @@
+            let cbsPointer = UnsafeMutablePointer<CBS>.allocate(capacity: 1)
+            defer { cbsPointer.deallocate() }
+            derRepresentation.regions.flatMap { $0 }.withUnsafeBufferPointer { buffer in
+                cbsPointer.pointee = CBS(data: buffer.baseAddress, len: buffer.count)
```
You can't do this: this propagates the pointer for the bytes out of the closure, which is forbidden. We need to use the CBS within this closure.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,319 @@
+            let cbsPointer = UnsafeMutablePointer<CBS>.allocate(capacity: 1)
+            defer { cbsPointer.deallocate() }
+            derRepresentation.regions.flatMap { $0 }.withUnsafeBufferPointer { buffer in
+                cbsPointer.pointee = CBS(data: buffer.baseAddress, len: buffer.count)
```
This is unsound: it escapes `buffer` from the closure, which is forbidden. You must use the CBS entirely within this closure.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,319 @@
+        /// Initialize a ML-DSA-65 public key from a PEM representation.
+        /// 
+        /// - Parameter pemRepresentation: The PEM representation of the public key.
+        public init(pemRepresentation: String) throws {
```
There is no PEM representation here.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,319 @@
+        }
+
+        /// The DER representation of the public key.
+        public var derRepresentation: Data {
```
Same note here, there's no DER representation.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,319 @@
+        /// The DER representation of the public key.
+        public var derRepresentation: Data {
+            get throws {
+                let cbbPointer = UnsafeMutablePointer<CBB>.allocate(capacity: 1)
```
We don't ever need to heap allocate the `CBB` structures, we can stack allocate them: `var cbb = CBB()`

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,319 @@
+                let cbbPointer = UnsafeMutablePointer<CBB>.allocate(capacity: 1)
+                defer { cbbPointer.deallocate() }
+
+                guard CCryptoBoringSSL_CBB_init(cbbPointer, MLDSA.PublicKey.bytesCount) == 1 else {
```
`CBB_init` can only return 0 on allocation failure, which we define as impossible. So we can force-unwrap this with a comment to that effect.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,319 @@
+        }
+
+        /// The PEM representation of the public key.
+        public var pemRepresentation: String {
```
Same note here, there's no PEM representation.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,319 @@
+        public func isValidSignature(_ signature: Signature, for data: some DataProtocol, context: [UInt8]? = nil) -> Bool {
+            return signature.withUnsafeBytes { signaturePtr in
+                let rc: CInt = data.regions.first!.withUnsafeBytes { dataPtr in
+                    let contextPointer = context?.withUnsafeBufferPointer { $0.baseAddress }
```
Same note here, you can't escape this pointer.

Lukasa commented on Sources/_CryptoExtras/Util/Error.swift:
```diff
@@ -15,3 +15,11 @@
 }
+
+/// Errors that can be thrown when working with ``MLDSA``.
+public enum CryptoMLDSAError: Error {
```
Let's not add a new error enum here, can we use our existing errors?

fpseverino replied:
Can we use `CryptoKitError.underlyingCoreCryptoError` when BoringSSL fails to generate a key or sign?

fpseverino replied:
Nevermind, I found `internalBoringSSLError()` 👍🏻

fpseverino commented:
Hi @Lukasa, thank you very much for the quick feedback and directions.

I think I fixed most of your requested changes, but I have some problems with the DER and PEM representations.
Unlike RSA, where there are clear BoringSSL methods for generating and parsing them, I can't find any for ML-DSA. Maybe there are some generic ones? Or a pure Swift implementation is possible/required?

Please let me know if the other changes I made are valid and how I could fix the PEM/DER issue.
Thanks again for the help!

fpseverino commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,428 @@
+        static let bytesCount = Backing.bytesCount
+
+        fileprivate final class Backing {
+            let pointer: UnsafeMutablePointer<MLDSA65_private_key>
```
So should I add a method to `PrivateKey.Backing` like this one?
```swift
func withUnsafePointer<T>(_ body: (UnsafePointer<MLDSA65_private_key>) throws -> T) rethrows -> T {
    try body(self.pointer)
}
```

fpseverino commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,428 @@
+            func signature(for data: some DataProtocol, context: [UInt8]? = nil) throws -> Signature {
+                let output = try Array<UInt8>(unsafeUninitializedCapacity: Signature.bytesCount) { bufferPtr, length in
+                    let result = data.regions.first!.withUnsafeBytes { dataPtr in
+                        if let context {
```
I'm having some trouble converting the optional `DataProtocol` to an optional `ContiguousBytes`

fpseverino replied:
Would this be ok?
```swift
let contextOptional: ContiguousBytes? = if let context { Data(context) } else { nil }
return contextOptional.withUnsafeBytes { contextPtr in
    CCryptoBoringSSL_MLDSA65_verify(
        self.pointer,
        signaturePtr.baseAddress,
        signaturePtr.count,
        dataPtr.baseAddress,
        dataPtr.count,
        contextPtr.baseAddress,
        contextPtr.count
    )
}
```

fpseverino commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,352 @@
+
+extension MLDSA {
+    /// A ML-DSA-65 private key.
+    public struct PrivateKey: Sendable {
```
BoringSSL doesn't offer a public API for the serialization of the private key

Lukasa replied:
That's extremely weird. Ah well, we should just write it ourselves. It's not a terribly complex format.

Lukasa replied:
Ok, I chatted with the BoringSSL folks briefly. It turns out the jury is out on exactly what the serialization format will be for these keys. As a result, let's follow their lead and for now only offer APIs to and from _seed_. We can add other formats when it becomes clear which ones are actually needed.

fpseverino replied:
So should I remove all the `rawRepresentation` inits and properties on private and public keys?

Lukasa replied:
Oh sorry, no, only on private keys. For public keys, the raw representation is what we need to use for now.

Lukasa replied:
As there's no "public key from seed" construction.

fpseverino replied:
Ok, so on public keys we need an initializer from raw and a property that returns the raw, whereas on the private key I have to add a property that returns the seed, right?

We could save the seed in the `PrivateKey` struct, as it is returned from the BoringSSL init.

Lukasa replied:
Yep, that's exactly right.

fpseverino commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,352 @@
+
+                self.pointer = UnsafeMutablePointer<MLDSA65_private_key>.allocate(capacity: 1)
+
+                try rawRepresentation.regions.flatMap { $0 }.withUnsafeBufferPointer { buffer in
```
When I rewrite it like
```swift
let bytes: ContiguousBytes = rawRepresentation.regions.count == 1 ? rawRepresentation.regions.first! : Array(rawRepresentation)
try bytes.withUnsafeBytes { buffer in
    var cbs = CBS(data: buffer.baseAddress, len: buffer.count)
    guard CCryptoBoringSSL_MLDSA65_parse_private_key(self.pointer, &cbs) == 1 else {
        throw CryptoKitError.internalBoringSSLError()
    }
}
```
an error on the CBS init appears, on `buffer.baseAddress`: `Cannot convert value of type 'UnsafeRawPointer?' to expected argument type 'UnsafePointer<UInt8>?' because initializer 'init(data:len:)' was not imported from C header`.

Changing it to `CBS(data: buffer.baseAddress?.assumingMemoryBound(to: UInt8.self), len: buffer.count)` seems to work, is it fine?

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,321 @@
+            /// Initialize a ML-DSA-65 private key from a random seed.
+            init() throws {
+                self.pointer = UnsafeMutablePointer<MLDSA65_private_key>.allocate(capacity: 1)
+                self.seedPointer = UnsafeMutablePointer<UInt8>.allocate(capacity: MLDSA.seedSizeInBytes)
```
I think we can make this a touch more efficient. We can store both of these fields in their useful formats. For the private key, we can just have a `var key: MLDSA65_private_key`, and then derive pointers as needed.

For the seed, we can use a `withUnsafeTemporaryAllocation` and then directly create a `Data` from it. `Data` is the format we need elsewhere anyway.

fpseverino replied:
This means that we don't need a `Backing` class?

Lukasa replied:
Technically it does, but we should keep it to avoid us abusing the power of `@_implementationOnly`.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,321 @@
+            /// 
+            /// - Throws: `CryptoKitError.incorrectKeySize` if the seed is not at least 32 bytes long.
+            init(seed: some DataProtocol) throws {
+                guard seed.count >= MLDSA.seedSizeInBytes else {
```
I can't see any reason to allow the seed to be too long.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,321 @@
+        static let bytesCount = Backing.bytesCount
+
+        fileprivate final class Backing {
+            private let pointer: UnsafeMutablePointer<MLDSA65_public_key>
```
Same note here, we can store the `MLDSA64_public_key` directly inline here.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,321 @@
+            var rawRepresentation: Data {
+                var cbb = CBB()
+                // The following BoringSSL functions can only fail on allocation failure, which we define as impossible.
+                CCryptoBoringSSL_CBB_init(&cbb, MLDSA.PublicKey.Backing.bytesCount)
```
This currently leaks, we need a `CBB_cleanup`.

Lukasa reviewed (commented):
Nice, we're getting a lot closer here!

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -16,9 +16,8 @@ import Crypto
-@_implementationOnly import CCryptoBoringSSLShims
 
-/// A lattice-based digital signature algorithm that provides security against quantum computing attacks.
+/// A module lattice-based digital signature algorithm that provides security against quantum computing attacks.
```
Do we want `module` here?

fpseverino replied:
I added it just to complete the ML-DSA acronym, maybe I could add an hyphen between it and `lattice` to make it clearer

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -68,25 +67,26 @@ extension MLDSA {
-                        self.pointer
-                    ) == 1 else {
-                        throw CryptoKitError.internalBoringSSLError()
+                (self.key, self.seed) = try withUnsafeTemporaryAllocation(of: MLDSA65_private_key.self, capacity: 1) { privateKeyPtr in
```
I don't think we need an `unsafeTemporaryAllocation`. We can just zero-initialize the value `self.key = .init()` and then take a pointer to it.

fpseverino replied:
Got it. I also tried to remove the seed's `unsafeTemporaryAllocation` by passing to the BoringSSL function an empty `[UInt8]` var and then initializing the `Data` object from it, but I encountered some seg faults in testing

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -101,24 +101,18 @@ extension MLDSA {
-
-                let seedBytes = Array(seed.prefix(MLDSA.seedSizeInBytes))
-                seedPointer.initialize(from: seedBytes, count: MLDSA.seedSizeInBytes)
+                self.key = try withUnsafeTemporaryAllocation(of: MLDSA65_private_key.self, capacity: 1) { privateKeyPtr in
```
Same note about zero-initialization here.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -203,12 +197,12 @@ extension MLDSA {
-                self.pointer = UnsafeMutablePointer<MLDSA65_public_key>.allocate(capacity: 1)
-                let _ = privateKeyBacking.withUnsafePointer { privateKeyPtr in
-                    CCryptoBoringSSL_MLDSA65_public_from_private(self.pointer, privateKeyPtr)
+                self.key = withUnsafeTemporaryAllocation(of: MLDSA65_public_key.self, capacity: 1) { publicKeyPtr in
```
Same note about zero-initialization here.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift:
```diff
@@ -0,0 +1,305 @@
+
+                guard CCryptoBoringSSL_MLDSA65_private_key_from_seed(
+                    &self.key,
+                    Array(seed.prefix(MLDSA.seedSizeInBytes)),
```
Let's not create an array here, nor should we use `prefix`. We know the seed size is right, and we can use `self.seed.withUnsafeBytes` to get the interior pointer there.

Lukasa reviewed (commented):
Cool, looking really good here. One note in the diff.

Lukasa reviewed (approved):
Great, I think this is in good shape. Let's get it merged!

Lukasa commented:
Just a few CI jobs to fix up. Missing license headers, and a need to run the cmake script in the `scripts` directory.

fpseverino commented:
The license headers are there, but I think the CI wants a `CONTRIBUTORS.txt`, whereas in this repo there is a `CONTRIBUTORS.md`

fpseverino commented:
Regarding the CMake script, if I run `sh scripts/update_cmakelists.sh` it removes all the libraries string from the `CMakeLists.txt` files. I suppose that's not the intended outcome.

Here's the output:
<details>

```
Finding source files (*.c *.swift) under /Users/francescopaoloseverino/Documents/GitHub/swift-crypto/Sources/CCryptoBoringSSL
scripts/update_cmakelists.sh: line 58: gfind: command not found

Updated /Users/francescopaoloseverino/Documents/GitHub/swift-crypto/Sources/CCryptoBoringSSL/CMakeLists.txt
Finding source files (*.c *.swift) under /Users/francescopaoloseverino/Documents/GitHub/swift-crypto/Sources/CCryptoBoringSSLShims
scripts/update_cmakelists.sh: line 58: gfind: command not found

Updated /Users/francescopaoloseverino/Documents/GitHub/swift-crypto/Sources/CCryptoBoringSSLShims/CMakeLists.txt
Finding source files (*.c *.swift) under /Users/francescopaoloseverino/Documents/GitHub/swift-crypto/Sources/CryptoBoringWrapper
scripts/update_cmakelists.sh: line 58: gfind: command not found

Updated /Users/francescopaoloseverino/Documents/GitHub/swift-crypto/Sources/CryptoBoringWrapper/CMakeLists.txt
Finding source files (*.c *.swift) under /Users/francescopaoloseverino/Documents/GitHub/swift-crypto/Sources/Crypto
scripts/update_cmakelists.sh: line 58: gfind: command not found

Updated /Users/francescopaoloseverino/Documents/GitHub/swift-crypto/Sources/Crypto/CMakeLists.txt
Finding source files (*.c *.swift) under /Users/francescopaoloseverino/Documents/GitHub/swift-crypto/Sources/_CryptoExtras
Excluding source paths (*/AES/*.swift) under /Users/francescopaoloseverino/Documents/GitHub/swift-crypto/Sources/_CryptoExtras
scripts/update_cmakelists.sh: line 58: gfind: command not found

Updated /Users/francescopaoloseverino/Documents/GitHub/swift-crypto/Sources/_CryptoExtras/CMakeLists.txt
Finding assembly files (.S) under /Users/francescopaoloseverino/Documents/GitHub/swift-crypto/Sources/CCryptoBoringSSL
scripts/update_cmakelists.sh: line 72: gfind: command not found
scripts/update_cmakelists.sh: line 73: gfind: command not found
scripts/update_cmakelists.sh: line 74: gfind: command not found
scripts/update_cmakelists.sh: line 75: gfind: command not found




Updated /Users/francescopaoloseverino/Documents/GitHub/swift-crypto/Sources/CCryptoBoringSSL/CMakeLists.txt
```

</details>

P.S. I do have the `find` command installed on my computer

Lukasa commented:
We expect GNU find. You can get it by running `brew install findutils`

fpseverino commented:
@Lukasa all done, now the jobs should pass

fpseverino commented:
I saw that the BoringSSL team added another parameter set to ML-KEM. Considering that they could do this for ML-DSA too, in order not to break the API should we put the current implementation inside a namespace that identifies the parameter set? If so I'm open to suggestions for the name, since the following (which is the most immediate one) won't work
```swift
enum MLDSA {
    enum 65 {}
    
    // Possible future ones
    enum 44 {}
    enum 87 {}
}
```

P.S. Each parameter set uses completely different C++ functions and structures

fpseverino commented:
Hi @Lukasa,
Thanks again for reviewing and approving this PR!
Just checking if it’s possible to get this merged when you have a moment.
Once this is merged, it would be great to start looking at reviewing #278 and #314.
Let me know if there’s anything else I need to do 🙂

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA65_boring.swift:
```diff
@@ -0,0 +1,320 @@
+
+extension MLDSA65 {
+    /// A ML-DSA-65 signature.
+    public struct Signature: Sendable, ContiguousBytes {
```
I don't think this type is pulling its weight, honestly. Right now there's only a "raw" ML-DSA signature format, and so we should treat this more like we treat Curve25519, where the output of the signing operation is just `Data`.

fpseverino commented:
There's an annoying thing in the `signature` and `isValidSignature` methods that I don't know how to solve:

I wanted to give a separate generic type to the `context` argument, like `Curve25519` (and now also `MLDSA65`) do for the `data` and `signature` arguments, but since it is `nil` by the default the compiler can't infer it.
The solution in place right now is to use the same generic type as `data`.

Lukasa commented on Sources/_CryptoExtras/MLDSA/MLDSA65_boring.swift:
```diff
@@ -0,0 +1,300 @@
+        }
+
+        /// The seed from which this private key was generated.
+        public var seed: Data {
```
Let's call this `seedRepresentation`.

Lukasa reviewed (commented):
Thanks for this, and sorry for the delay in review! Generally I'm really happy with how this looks. I've left a few notes in the diff.

fpseverino commented on Sources/_CryptoExtras/MLDSA/MLDSA65_boring.swift:
```diff
@@ -0,0 +1,300 @@
+        static let byteCount = Backing.byteCount
+
+        fileprivate final class Backing {
+            var key: MLDSA65_private_key
```
I have to make this `fileprivate` because it's used in the public key backing

Lukasa reviewed (commented):
OK, two very minor nits and then I think we're good to go. I'll also kick the tests off here so we have that information.

Lukasa reviewed (approved):
Ok, this looks really nice. Thanks for your hard work here @fpseverino! :sparkles:

fpseverino commented:
Thank you @Lukasa for all the help and guidance you have given me!
